### ISSUE TITLE
tf.nn.ctc_loss not returning expected value

### ISSUE THREAD

**Issue by lriuui0x0** [tensorflow/tensorflow#41353]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below):  v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.7.7
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No

**Describe the current behavior**

I run the following code:

```
import tensorflow as tf

label = [1, 2, 1]
logits = [[0.0, 1.0, 0.0],
          [0.0, 0.0, 1.0],
          [0.0, 1.0, 0.0]]
labels_length = 3
logits_length = 3

labels_tensor = tf.convert_to_tensor([label], dtype=tf.int32)
logits_tensor = tf.convert_to_tensor([logits], dtype=tf.float32)
labels_length_tensor = tf.convert_to_tensor([labels_length], dtype=tf.int32)
logits_length_tensor = tf.convert_to_tensor([logits_length], dtype=tf.int32)

loss = tf.nn.ctc_loss(labels_tensor, logits_tensor, labels_length_tensor, logits_length_tensor, logits_time_major=False)
print(loss.numpy()[0])
```

Basically there're 3 timestamps, and there're two characters `a, b` and the GT is `aba` (`[1, 2, 1]`) which only has one representation in CTC. And I'm making the logits have the probability of 1 for `aba` only. Therefore I expect the loss is 0, however the loss is some random value 1.6543342.

Either there's a bug, or I seriously misunderstand the documentation - in which case I think the [documentation on this topic](https://www.tensorflow.org/api_docs/python/tf/nn/ctc_loss) isn't very good. It would be better to have some example, as suggested in another issue https://github.com/tensorflow/tensorflow/issues/35063.

**Describe the expected behavior**

Expect the loss to be 0.

**Comment by ravikyram** [comment]
I have tried in colab with TF versions 2.2, 2.3-rc1,nightly versions(`2.4.0-dev20200712`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/4ae1e375a30ac4a1d887525a71b50611/untitled119.ipynb).Thanks!

**Comment by WindQAQ** [comment]
Hi @lriuui0x0 , I think the value is not expected to zero in the implementation. If we follow the fomula in the paper, the neg log probability is of course zero (see last two cell in the colab link). However, in most of frameworks, in order for numerical stability, the product of probability will become the addition of log probability, and thus, the input probability will be transformed via log softmax first, and do the computation in logarithm domain. You can see the such transformation in [this line](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/ctc_ops.py#L651).

In this kind of implementation (see the last cell in the colab link), the produced loss value is indeed 1.6543342. Note that pytorch (and cudnn backend) also follows the same implementation for numerical stability

https://colab.research.google.com/drive/1DofmAw8wgsYYN166n-s80PPKA092U6F6?usp=sharing
https://www.cs.toronto.edu/~graves/icml_2006.pdf

**Comment by gowthamkpr** [comment]
@lriuui0x0 As mentioned in the above comment, this is an expected behavior.

**Comment by lriuui0x0** [comment]
@WindQAQ @gowthamkpr Thanks for the response. It's quite a big off from the actual value `0`. Is it due to `np.logaddexp` being inaccurate?

**Comment by WindQAQ** [comment]
> @WindQAQ @gowthamkpr Thanks for the response. The value is quite off from the actual value `0`. Is it due to `np.logaddexp` being inaccurate?

No, it is due to log softmax will assign some small probability to zero probability. For example, in your case [0.0, 1.0, 0.0] after log softmax will become [-1.55144471 -0.55144471 -1.55144471]. Obviously, if you take exponent on it (convert it back to probability), they are all non-zero. If you use this probability (`e^[-1.55144471 -0.55144471 -1.55144471]`) to compute ctc loss, you can also obtain 1.6543342.

**Comment by lriuui0x0** [comment]
@WindQAQ OK so it's because of the way `tf.nn.log_softmax` is impelemented, it results in `[-1.55144471 -0.55144471 -1.55144471]` instead of `[-inf 0 -inf]`, which is mathematically more correct, but is numerically unstable / slower.

OK that makes sense then, do you know any reference on the theory behind why log softmax is implemented in that way?

**Comment by WindQAQ** [comment]
> @WindQAQ OK so it's because of the way `tf.nn.log_softmax` is impelemented, it results in `[-1.55144471 -0.55144471 -1.55144471]` instead of `[-inf 0 -inf]`, which is mathematically more correct, but is numerically unstable / slower.
> 
> OK that makes sense then, do you know any reference on the theory behind why log softmax is implemented in that way?

Umm, I am not sure about that, but it's not about how `log_softmax` is implemented, it is implemented as described in its (take softmax and then take log).

**Comment by lriuui0x0** [comment]
@WindQAQ If it's really implementing softmax then `element-wise` log, why would `[0, 1, 0]` results in `[-1.55144471, -0.55144471, -1.55144471]`?

**Comment by WindQAQ** [comment]
Denote `z = [0, 1, 0]`.
Let's first compute the numerator of softmax, that is, `e^{z_i}`.
We can have `e^z = [1, e, 1]`. The denominator is then `sum_{i=0}^2 e^{z_i}` = 2 + e. Thus, the softmax value is `softmax(z) = [1 / (2 + e), e / (2 + e), 1 / (2 + e)]`. Finally, taking the log on softmax yields `log_softmax(z) = [-1.55144471, -0.55144471, -1.55144471]`.

**Comment by lriuui0x0** [comment]
@WindQAQ Oh yes, I forgot it's calculating softmax once again on top of softmax result. Thanks, we can close the issue now!
